### PR TITLE
Auto create timeseries subscriptions

### PR DESCRIPTION
**Pull request by wbevans**
This pull request primarily focuses on the automation of subscription creation and the simplification of subscription configuration in the CDF Fabric Replicator program. The changes remove the need for manual subscription setup and replace the list of subscriptions with a single subscription object. 

Key Changes:

1. **Automated Subscription Creation:** The time series replicator now automatically creates the subscription based on the values in the configuration file. This removes the need for manual subscription setup using the Python SDK, simplifying the process.

2. **Simplified Subscription Configuration:** The `subscriptions` list in the configuration files has been replaced with a single `subscription` object. This change simplifies the configuration and reflects the shift towards automated subscription creation. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L91-R51) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L122-R86) [[3]](diffhunk://#diff-856023380e68b3ef1d677e2656c3b6181f568d8b241e68195aa65c59998ee4e2L18-R19) [[4]](diffhunk://#diff-856023380e68b3ef1d677e2656c3b6181f568d8b241e68195aa65c59998ee4e2L55-R56) [[5]](diffhunk://#diff-696623dffe1c8028d784a78b7dbfa60d8d9fa23454d2fbb85cfbd814756126a3L31-R34) [[6]](diffhunk://#diff-2298d1e12fe0cb078bf8cae0c2b3c74eed6e0d2c19c35d8e4d799c31ae672af8L30-R35) [[7]](diffhunk://#diff-40036ccfae2c64630df36fca137f88608efd30b837b750a1cacde8130d4417d1L29-R32)

3. **Codebase Updates:** Changes were made to the `cdf_fabric_replicator/time_series.py`, `cdf_fabric_replicator/config.py`, and test files to accommodate the new subscription creation and configuration process. These changes include the addition of a `create_subscription` method and updates to log messages and method parameters. [[1]](diffhunk://#diff-21a04836ddcc3134221a000e80759bac758f662ee1d35da9b6fa18ecef6f4e5eL12-R25) [[2]](diffhunk://#diff-21a04836ddcc3134221a000e80759bac758f662ee1d35da9b6fa18ecef6f4e5eL48-L58) [[3]](diffhunk://#diff-21a04836ddcc3134221a000e80759bac758f662ee1d35da9b6fa18ecef6f4e5eL78-R94) [[4]](diffhunk://#diff-21a04836ddcc3134221a000e80759bac758f662ee1d35da9b6fa18ecef6f4e5eR270-R282) [[5]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL14-L17) [[6]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL89-L101) [[7]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL262-R251) [[8]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL280-R268) [[9]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL18) [[10]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL51-R50) [[11]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL65) [[12]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bR94-R95) [[13]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL107-R107)

4. **Test Updates:** The test files were updated to reflect the new subscription creation and configuration process. These changes include the removal of the `create_subscription_in_cdf` method and updates to method parameters and assertions. [[1]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL14-L17) [[2]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL89-L101) [[3]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL262-R251) [[4]](diffhunk://#diff-9857b8ba3a329c52b4cf53e9533b4d52ef2b0c01f628f50a98f93b009c0470dcL280-R268) [[5]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL18) [[6]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL51-R50) [[7]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL65) [[8]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bR94-R95) [[9]](diffhunk://#diff-c86d641053fcfce65db697b5b95258100a450af69245534d7d196557f812a64bL107-R107) [[10]](diffhunk://#diff-4f8ce20ee43f8c887996be4b70de530a19ba527eafc86867bfc96f97b3ca3888L210-R210) 

5. **Configuration File Updates:** The configuration files were updated to replace the `subscriptions` list with a single `subscription` object. This change simplifies the configuration and reflects the shift towards automated subscription creation. [[1]](diffhunk://#diff-696623dffe1c8028d784a78b7dbfa60d8d9fa23454d2fbb85cfbd814756126a3L31-R34) [[2]](diffhunk://#diff-2298d1e12fe0cb078bf8cae0c2b3c74eed6e0d2c19c35d8e4d799c31ae672af8L30-R35) [[3]](diffhunk://#diff-40036ccfae2c64630df36fca137f88608efd30b837b750a1cacde8130d4417d1L29-R32)

Testing:

The integration tests were updated to account for subscription creation during the run of the replicator, so they should run with no special actions. However, when testing the replicator itself, it might be useful to delete the subscription manually before running the replicator, as the creation code is only executed on the first run. This code will delete a subscription:

```
client.time_series.subscriptions.delete("cdf_fabric_replicator_sub")
```

### CODE REVIEW

File: cdf_fabric_replicator/time_series.py [hunk 1]
```diff
@@ -75,13 +76,24 @@ def run(self) -> None:
...
 
     def process_subscriptions(self) -> None:
-        for subscription in self.config.subscriptions:
-            for partition in subscription.partitions:
-                with ThreadPoolExecutor() as executor:
-                    future = executor.submit(
-                        self.process_partition, subscription, partition
-                    )
-                    logging.debug(future.result())
+        if (
```
Comment (xtophs) [inline cdf_fabric_replicator/time_series.py:79]: What about unit tests for this new functionality?

File: cdf_fabric_replicator/time_series.py [hunk 2]
```diff
@@ -75,13 +76,24 @@ def run(self) -> None:
...
+            try:
+                self.create_subscription(self.config.subscription.num_partitions)
+            except CogniteAPIError as e:
+                logging.error(f"Error creating subscription: {e}")
+                raise e
+
+        for partition in range(0, self.config.subscription.num_partitions):
+            with ThreadPoolExecutor() as executor:
+                future = executor.submit(
+                    self.process_partition, self.config.subscription, partition
```
Comment (xtophs) [inline cdf_fabric_replicator/time_series.py:94]: It appears that a failure writing a data point to the lakehouse would take down this entire thread without logging the failure or ensuring a retry.

File: cdf_fabric_replicator/time_series.py [hunk 3]
```diff
@@ -255,3 +267,16 @@ def get_token(self) -> str:
...
+            f"Subscription {self.config.subscription.external_id} not found. Creating subscription..."
+        )
+
+        sub = DataPointSubscriptionWrite(
+            external_id=self.config.subscription.external_id,
+            name=self.config.subscription.name,
+            partition_count=num_partitions,
+            filter=flt.Exists(TimeSeriesProperty.external_id),
+        )
+        self.cognite_client.time_series.subscriptions.create(sub)
```
Comment (xtophs) [inline cdf_fabric_replicator/time_series.py:282]: What about logging success here?

File: cdf_fabric_replicator/time_series.py [hunk 4]
```diff
@@ -75,13 +76,24 @@ def run(self) -> None:
...
+            is None
+        ):
+            try:
+                self.create_subscription(self.config.subscription.num_partitions)
+            except CogniteAPIError as e:
+                logging.error(f"Error creating subscription: {e}")
+                raise e
+
+        for partition in range(0, self.config.subscription.num_partitions):
+            with ThreadPoolExecutor() as executor:
```
Comment (xtophs) [inline cdf_fabric_replicator/time_series.py:92]: Why process this in another thread?
Could this lead to a race conditions with the cursor being read and written to the shared state store? 
Do we need to ensure rows in partitions are inserted into the lakehouse in the order they were received from CDF?

File: tests/integration/test_timeseries_replicator.py [hunk 5]
```diff
@@ -93,6 +91,8 @@ def test_timeseries_data_integration_service(
     azure_credential,
     remote_state_store,
 ):
+    # Run replicator before pushing data points in order to setup subscription
+    run_replicator(test_replicator)
```
Comment (xtophs) [inline tests/integration/test_timeseries_replicator.py:95]: Do we need any assertions (or new tests) to assert proper creation of subscriptions and proper handling of edge conditions, e.g. multiple subscriptions existing?

File: cdf_fabric_replicator/time_series.py (hunk 1 above)
Comment (wbevans) [reply cdf_fabric_replicator/time_series.py:79]: Added unit tests

File: cdf_fabric_replicator/time_series.py (hunk 3 above)
Comment (wbevans) [reply cdf_fabric_replicator/time_series.py:286]: Added success message

File: cdf_fabric_replicator/time_series.py (hunk 4 above)
Comment (wbevans) [reply cdf_fabric_replicator/time_series.py]: The original intent was for performance, however, it appears that there could be race conditions with the use of class variables such as update_queue. In addition, if partitions were completed out of order, if could mess up the state store for retry if one of the earlier partitions fails after a later one succeeds and gets marked in the state store. Removing the threading logic until it is identified that more performance is needed at which time it can be re-added in a more thread-safe way.

File: cdf_fabric_replicator/time_series.py (hunk 2 above)
Comment (wbevans) [reply cdf_fabric_replicator/time_series.py]: See the other comments for information on threading, which has been removed. Added try / catch logic to log errors for writing to Lakehouse and iterating over timeseries. Additional error handling and logic is scheduled for implementation in the upcoming story # 37362 ( As a platform operator, I want a robust replication for the Fabric Connector ).

File: tests/integration/test_timeseries_replicator.py (hunk 5 above)
Comment (wbevans) [reply tests/integration/test_timeseries_replicator.py:96]: Added assert_subscription_created_in_cdf method

File: config-petro-tech.yaml [hunk 6]
```diff
@@ -28,12 +28,12 @@ extractor:
...
-  - external_id: gauge-ts
-    partitions:
-        - 0
-    lakehouse_abfss_path_dps: abfss://a0fd605a-3a69-401e-a586-bee3bf2c8127@onelake.dfs.fabric.microsoft.com/40b75c4d-5ddd-47ff-bd81-8faadec1804b/Tables/Timeseries
-    lakehouse_abfss_path_ts: abfss://a0fd605a-3a69-401e-a586-bee3bf2c8127@onelake.dfs.fabric.microsoft.com/40b75c4d-5ddd-47ff-bd81-8faadec1804b/Tables/TimeseriesMetadata
+subscription:
+  external_id: gauge-ts
+  name: Gauge Timeseries
+  num_partitions: 1
+  lakehouse_abfss_path_dps: abfss://a0fd605a-3a69-401e-a586-bee3bf2c8127@onelake.dfs.fabric.microsoft.com/40b75c4d-5ddd-47ff-bd81-8faadec1804b/Tables/Timeseries
```
Comment (nicholasperkins1992) [inline config-petro-tech.yaml:35]: I know you did not introduce this file, but does it make sense to have a config with filled out values

File: cdf_fabric_replicator/time_series.py [hunk 7]
```diff
@@ -241,17 +253,34 @@ def write_pd_to_deltalake(
...
+    ) -> None:
+        logging.debug(f"Subscription {external_id} not found. Creating subscription...")
+
+        sub = DataPointSubscriptionWrite(
+            external_id=external_id,
+            name=name,
+            partition_count=num_partitions,
+            filter=flt.Exists(TimeSeriesProperty.external_id),
+        )
+        self.cognite_client.time_series.subscriptions.create(sub)
```
Comment (nicholasperkins1992) [inline cdf_fabric_replicator/time_series.py:285]: Thinking about single responsibility, does it make sense to have all methods for creating subscription in the time series.py, might be worth some thought so see if we need a subscription.py

File: config-petro-tech.yaml (hunk 6 above)
Comment (wbevans) [reply config-petro-tech.yaml]: Updated to use env variables

File: cdf_fabric_replicator/time_series.py (hunk 7 above)
Comment (wbevans) [reply cdf_fabric_replicator/time_series.py]: Moved create_subscription to subscription.py

File: README.md [hunk 8]
```diff
@@ -119,13 +79,13 @@ extractor:
...
-    - external_id: ts-subscription
-      partitions:
-          - 0
-      lakehouse_abfss_path_dps: ${LAKEHOUSE_ABFSS_PREFIX}/Tables/${DPS_TABLE_NAME}
-      lakehouse_abfss_path_ts: ${LAKEHOUSE_ABFSS_PREFIX}/Tables/${TS_TABLE_NAME}
+# subscription to stream (will be automatically created with the given properties)
+subscription:
+    external_id: ts-subscription
+    name: My Subscription
+    num_partitions: 1
```
[the remark this answers is not in the document]
Comment (muradsater) [reply README.md:86]: @wbevans You are removing the possibility to split processing of a subscription on multiple workers, that's the whole point of having multiple partitions. My suggestion is to keep the old config format and only auto-create subscription if you set it to process partition 0. Then we can support more advanced setups where the simple case is auto-created, but more advanced setups are configured by the user.

You also removed the functionality to process multiple subscription, could we keep that functionality and do auto-create if there is one subscription defined?

**Review by muradsater — CHANGES_REQUESTED** [review]
@wbevans You are removing the possibility to split processing of a subscription on multiple workers, that's the whole point of having multiple partitions. My suggestion is to keep the old config format and only auto-create subscription if you set it to process partition 0. Then we can support more advanced setups where the simple case is auto-created, but more advanced setups are configured by the user.

You also removed the functionality to process multiple subscription, could we keep that functionality and do auto-create if there is one subscription defined?

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (wbevans) [reply README.md:86]: The idea of using one subscription was to simplify the configuration for users. I took your suggestion and re-enabled the multiple subscriptions and only auto-create if there is a single subscription that does not exists and that subscription only contains a single 0 partition.

As for the Threading, we identified a few issues with some of the threading code that exists now. However, for the sake of simplicity, I will leave the threading as it was originally for this PR and submit an additional PR for improving the threading.

**Review by muradsater — APPROVED** [review]
